Pull request for code review

juliusval opened the pull request.

juliusval committed:
Added job service, repo, controller, model with basic methods

juliusval committed:
Fixed formatting

juliusval committed:
Fixed model, getters and setters

juliusval committed:
Minor changes

roguefaction committed:
Merge branch 'heroku-db-setup' of https://github.com/roguefaction/good-deeds-be into master

# Conflicts:
#	src/main/resources/application.properties

juliusval committed:
Merge remote-tracking branch 'origin/master' into JobFunctionality

# Conflicts:
#	src/main/java/com/example/gooddeedsbe/GoodDeedsBeApplication.java

roguefaction committed:
added exceptions

roguefaction committed:
added JobValidator class

juliusval committed:
Changed method accessibility to public in jobValidator class

roguefaction committed:
added cors

juliusval committed:
Updated job model

juliusval committed:
Added validation calls to service class

juliusval committed:
Added exception handler to controller

juliusval committed:
Fixed job validator class

juliusval committed:
Added unit tests for job validator

juliusval committed:
Merge branch 'master' into JobValidation

juliusval committed:
Added put method for post editing, not finished

juliusval committed:
Fixed editJob service method

juliusval committed:
Small fix

juliusval committed:
Added get job by id method in controller

juliusval committed:
getAllJobs returns with order by id

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+
+@CrossOrigin(origins = "*")
+@RestController
+public class JobController {
```
Would recommend renaming everything Job -> Deed. Job has a different connotation in back end, so reading this feels bit weird.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+
+import java.util.List;
+
+@CrossOrigin(origins = "*")
```
Current `CrossOrigin` configuration allows your API to be called from any other site. This should be restricted to only what you want your API be called from.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+import com.example.gooddeedsbe.service.JobService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
```
This was probably done by an IDE you're using, but in general you should avoid importing everything, as in `bind.annotation.*`, since a lot of different libs have stuff named the same, it may lead to errors by using wrong lib.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+        this.jobService = jobService;
+    }
+
+    @GetMapping(value = "/job")
```
Endpoint should reflect a list will be returned instead of single object.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+
+    @GetMapping(value = "/job/{id}")
+    public Job getJobById(@PathVariable int id) {
+        try{
```
Formatting

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+    public Job getJobById(@PathVariable int id) {
+        try{
+            return jobService.getJobById(id);
+        }catch (InvalidFieldException ex) {
```
Wrong exception cought, there is no filed here.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+            return jobService.getJobById(id);
+        }catch (InvalidFieldException ex) {
+            throw new ResponseStatusException(
+                    HttpStatus.BAD_REQUEST, ex.getMessage(), ex);
```
Wrong Http status. The method should result in 404 (not found) when the deed couldn't be found by id.
Also you probably don't need to pass exception it self and the message should be something meaningful.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+            return jobService.createJob(job);
+        }catch (InvalidFieldException ex){
+            throw new ResponseStatusException(
+                    HttpStatus.BAD_REQUEST, ex.getMessage(), ex);
```
Probably don't need to pass `ex`, same everywhere else in this file.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+        try {
+            return jobService.createJob(job);
+        }catch (InvalidFieldException ex){
+            throw new ResponseStatusException(
```
Throwing exceptions could be moved to service, while controller would only have them in method signature.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @DeleteMapping(value = "/job/{id}")
+    public void deleteJob(@PathVariable int id) {
```
The method for delete should return response status. Http OK if deletion succeeded, not found if deed wasn't found, etc.

enorkus commented on src/main/java/com/example/gooddeedsbe/model/Job.java:
```diff
@@ -0,0 +1,132 @@
+
+@Entity
+@Table(name = "job")
+public class Job implements Serializable {
```
Remove setters, remember what I thought you about immutability.

enorkus commented on src/main/java/com/example/gooddeedsbe/service/JobServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+        if(job.isPresent()){
+            return job.get();
+        } else {
+            throw new InvalidFieldException("Job not found with given ID");
```
Wrong exception.

enorkus commented on src/main/java/com/example/gooddeedsbe/service/JobServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public Job createJob(Job job) throws InvalidFieldException {
```
Catch the exception here and re-throw a HTTP response exception.

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobHelper.java:
```diff
@@ -0,0 +1,17 @@
+import com.example.gooddeedsbe.exceptions.InvalidFieldException;
+import com.example.gooddeedsbe.model.Job;
+
+public class JobHelper {
```
Not a helper, should be renamed to Validator.

enorkus replied:
The class only makes sense if you have different classes for different validators, otherwise this could as well be deleted and `validateJob(...)` method moved to the `JobValidator` class.

enorkus commented on src/main/java/com/example/gooddeedsbe/controller/JobController.java:
```diff
@@ -0,0 +1,64 @@
+package com.example.gooddeedsbe.controller;
```
Remove the `be` from packages. It's all backend in here :)

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobValidator.java:
```diff
@@ -0,0 +1,157 @@
+
+    private static final String EMAIL_REGEX = "^[\\w-\\+]+(\\.[\\w]+)*@[\\w-]+(\\.[\\w]+)*(\\.[a-z]{2,})$";
+    private static final String PHONE_NUMBER_REGEX = "^\\+370[0-9]{8}";
+    private static final String HASHTAG_REGEX_1 = ".*,$";
```
2 regexes for hashtag?

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobValidator.java:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        if(title.trim().length() == 0){
+            throw new TitleInvalidException("Title can't be whitespace");
```
There is no need for whitespace message. It can fall under the `Title can't be empty` message.

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobValidator.java:
```diff
@@ -0,0 +1,157 @@
+
+
+    public static void validateOrganization(String organization) throws OrganizationInvalidException {
+        if(StringValidator.checkIfNullOrWhitespace(organization))
```
I'd advise moving `checkIfNullOrWhitespace()` method here. Separate class is not needed. Furthermore, check out apache commons libraries for already existing solutions to check if string is null, empty or whitespace. An example could be StringUtils. No need to reinvent already existing and tested solutions.

enorkus commented on src/main/java/com/example/gooddeedsbe/exceptions/CityInvalidException.java:
```diff
@@ -0,0 +1,8 @@
+package com.example.gooddeedsbe.exceptions;
+
+public class CityInvalidException extends InvalidFieldException {
```
I'd suggest instead of all these exceptions, having a single `ValidationException` class which would have a validation message, which could be passed through constructor. All these exceptions adds unnecessary complexity and classes.

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobValidator.java:
```diff
@@ -0,0 +1,157 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JobValidator {
```
If you had different validators for different things, this one could instantiate and call them. You would pass whole job object to the validator and validate different attributes one by one, using specific validators. Try to define general validators applicable for multiple fields and specific ones. Like MandatoryValueValidator, InputLengthValidator, PhoneNumberValidator etc. etc.

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/JobValidator.java:
```diff
@@ -0,0 +1,157 @@
+
+    public static void validateCity(String city) throws CityInvalidException {
+
+        if(city == null){
```
This is repeated all over. Suggestion here is to create a `MandatoryValueValidator` and validate mandatory values using it.

enorkus commented on src/main/java/com/example/gooddeedsbe/utils/StringValidator.java:
```diff
@@ -0,0 +1,12 @@
+package com.example.gooddeedsbe.utils;
+
+public class StringValidator {
```
Unnecessary class. Can be moved under validator.

enorkus commented on src/test/java/com/example/gooddeedsbe/JobValidatorTests.java:
```diff
@@ -0,0 +1,255 @@
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
Remove as you don't need app context for unit tests.

enorkus commented on src/test/java/com/example/gooddeedsbe/JobValidatorTests.java:
```diff
@@ -0,0 +1,255 @@
+public class JobValidatorTests {
+
+	@Test
+	public void contextLoads() {
```
Remove, this does nothing.

enorkus commented on src/test/java/com/example/gooddeedsbe/JobValidatorTests.java:
```diff
@@ -0,0 +1,255 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class JobValidatorTests {
```
Unit tests for specific class should reside in separate test class. For example JobValidator - JobValidatorTest, JobService - JobServiceTest.

enorkus commented on src/main/java/com/example/gooddeedsbe/service/JobServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+            newJob.setId(id);
+            return jobRepository.save(newJob);
+        } else {
+            throw new InvalidFieldException("Job not found with given ID");
```
Wrong exception.

roguefaction committed:
Renamed Jobs -> Deeds
Fixed imports so classes only import required libs

roguefaction committed:
Changed list endpoint to reflect that a list will be returned

roguefaction committed:
changed cors annotation to only accept requests from the front end app

roguefaction committed:
fixed errors made while refactoring names
added new exception for incorrect id handling
fixed some error handling in DeedController for better clarity

roguefaction committed:
removed setters from Deed object
added a builder pattern to Deed object
changed editDeed in DeedServiceImpl to comply with the removed setters

roguefaction committed:
added apache commons libraries, removed String validator class
moved validateDeed to DeedValidator
changed base package name from gooddeedsbe to gooddeeds
some logical fixes in DeedValidator

roguefaction committed:
removed unnecessary imports and annotations from DeedValidatorTests

roguefaction committed:
moved all of the exception throwing and catching inside DeedService
added http status messages for deleteDeed (renamed from deleteJob)

roguefaction committed:
deleted unused exception
added vars instead of logins to application properties

roguefaction committed:
reverted cross origin change

roguefaction committed:
field now validate lithuanian letters correctly

jnr25 commented on src/main/resources/application.properties:
```diff
@@ -1,8 +1,15 @@
-spring.datasource.driver-class-name=org.h2.Driver
+spring.datasource.url= jdbc:postgresql://ec2-54-235-67-106.compute-1.amazonaws.com:5432/d15qhr7mvcq62a
+spring.datasource.username=mertikiflqlmei
+spring.datasource.password=be233e24136df7e8b1bc6cd1b34decb96113035e68fb5b02d9112bada841555f
```
pakeitėt slaptažodį?

roguefaction replied:
pakeičiau credentials bet nežinau kaip ištrinti iš commit history

roguefaction committed:
added new mandatory date field to deed model along with all required validations and tests
fixed lithuanian letter checking regex

roguefaction committed:
added a new endpoint to get 3 most upcoming deeds

roguefaction committed:
renamed date to deedDate for front end compatibility

roguefaction committed:
quick fix

juliusval committed:
Revert "quick fix"

This reverts commit ccf607ff

juliusval committed:
Revert "renamed date to deedDate for front end compatibility"

This reverts commit 4c52cbf6

roguefaction committed:
added new endpoint methods for all upcoming deeds in service

roguefaction committed:
added a new enpoint for all deeds '/allupcomingdeeds'

roguefaction committed:
fixed upcoming deed endpoints to include today's deeds

roguefaction committed:
created ApplicationUser model

roguefaction committed:
Added SecurityConfiguration

roguefaction committed:
added UserController for registering new users

roguefaction committed:
Created ApplicationUserRepository

roguefaction committed:
Created security package

roguefaction committed:
Added a few services

roguefaction committed:
quick fix for getting lists for unregistered users

roguefaction committed:
Added validation for user creation

roguefaction committed:
Added tests for password validation

roguefaction committed:
lower auth times for testing

roguefaction committed:
testing authorization access

roguefaction committed:
testing fixes

roguefaction committed:
Update application.properties

roguefaction committed:
/login now returns JWT
/info endpoint for user info

roguefaction committed:
Merge remote-tracking branch 'origin/master'

roguefaction committed:
added cors support for user controller

juliusval committed:
Made email field unique

juliusval committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/com/example/gooddeeds/controller/UserController.java

juliusval committed:
Added user to deed relationship, get user deeds method

juliusval committed:
Changed deed edit and deletion to fetch appropriate user from token (else throw exc)
